agent/websockets/connection.go

```diff
@@ -252,3 +264,60 @@ func (conn *Connection) ReadServerMessages() ([]interface{}, error) {
 		return nil, nil
 	}
 }
+
+// InjectedHeader represents a single header which is injected into a websocket message.
+type InjectedHeader struct {
+	// Header is the header to inject from.
+	Header string `json:"header,omitempty"`
+	// Path is a period-delimited string where each element refers to the child
+	// element of the previous element. For example, for the injection path `a.b.c`, it will
+	// inject the header value into the nested JSON objects for key c under key b under key a
+	// under the parent: `{"a": {"b": {"c": {"<key>": "<injected_header_value>"}}}}`
+	Path []string `json:"path,omitempty"`
+	// Key is the key for which to inject the value as.
+	Key string `json:"key,omitempty"`
+}
+
+// injectWebsocketMessage injects a shim header value into a single websocket message in-place.
+// Returns a pointer to a new copy of the struct on success.
+func injectWebsocketMessage(msg *message, injectionPath []string, injectionValues map[string]string) (*message, error) {
```

Comment:
nit: name it InjectionPathParent and InjectionKeyValues 

In following example: path == store.book.category value== reference (this seems to be the convention AFAIK) 
{
    "store": {
        "book": [
            {
                "category": "reference",
                "author": "Nigel Rees",
                "title": "Sayings of the Century",
                "price": 8.95
            },
        ]
    }
}
while your definition is:
path = store.book and value=map[string]string  {
                "category": "reference",
                "author": "Nigel Rees",
                "title": "Sayings of the Century",
                "price": 8.95
            },

Reply:
In the case of headers, the way we are doing it is fine because a header is expected to be unique.

However, the important thing to remember is that this injection needs to match the format of the CDAP websocket message, so it's tricky to change the format up too much, especially if we're trying to inject all header values and not just a specific header.